packages/phone-number-privacy/signer/src/key-management/key-provider-base.ts

```diff
@@ -1,28 +1,58 @@
+import { KeyName } from '@celo/phone-number-privacy-common'
+import config from '../config'
+export interface Key {
+  name: KeyName
+  version: number
+}
 export interface KeyProvider {
-  fetchPrivateKeyFromStore: () => Promise<void>
-  getPrivateKey: () => string
+  fetchPrivateKeyFromStore: (key: Key) => Promise<void>
+  getPrivateKey: (key: Key) => string
+  getPrivateKeyOrFetchFromStore: (key: Key) => Promise<string>
 }
 
 const PRIVATE_KEY_SIZE = 72
 
 export abstract class KeyProviderBase implements KeyProvider {
-  protected privateKey: string | null = null
+  protected privateKeys: Map<Key, string> = new Map()
 
-  public getPrivateKey() {
-    if (!this.privateKey) {
-      throw new Error('Private key is empty, provider not properly initialized')
+  public getPrivateKey(key: Key) {
+    const privateKey = this.privateKeys.get(key)
+    if (!privateKey) {
+      throw new Error(`Private key is unavailable: ${JSON.stringify(key)}`)
     }
+    return privateKey
+  }
 
-    return this.privateKey
+  public async getPrivateKeyOrFetchFromStore(key: Key): Promise<string> {
+    if (key.version < 1 || key.version > 10) {
```

Comment:
Unless there is a particular reason for the restriction, I would not include it. If we implement proactive resharing, we would see more than 10 key versions. Without it, requesting a key version too high seems harmless to me, but I could be missing something.
```suggestion
    if (key.version < 1) {
```